Switch to using ES Modules

jonkoops opened the pull request:
Closes #238

jonkoops commented on README.md:
````diff
@@ -130,23 +128,20 @@ This version is slower (about 5x) so it is offered as an opt-in.
 
 ```js
-var classNames = require('classnames/dedupe');
+import classNames from 'classnames/dedupe.js';
````
One of the 'downsides' of using the ES module syntax is that it is [mandatory to provide a file extension](https://nodejs.org/api/esm.html#esm_mandatory_file_extensions) for files.

remcohaszing replied:
This can be worked around by adding [entry points](https://nodejs.org/api/packages.html#packages_package_entry_points) in `package.json`.

```js
{
  // …
  "exports": {
    ".": "./index.js",
    "./bind": "./bind.js",
    "./dedupe": "./dedupe.js"
  },
  // …
}
```

Although personally I prefer using named exports.

```js
import { classNames, bind, dedupe } from 'classnames';
```

jonkoops replied:
Nice, I'll add the entry points to the package. We'll have to discuss the named exports, they could be added as a feature later also.

jonkoops replied:
Ok, I've added a commit for the entry points. I'm resolving this one.

jonkoops commented:
The browser tests are also no longer functional as Browserify does not know how to handle the new syntax.

jonkoops commented on benchmarks/run.js:
```diff
@@ -1,17 +1,20 @@
-	var npm = require('classnames');
-	var npmDedupe = require('classnames/dedupe');
-	var npmPackage = require('./node_modules/classnames/package.json');
+	var npm = (await import('classnames')).default;
+	var npmDedupe = (await import('classnames/dedupe.js')).default;
+	var npmPackage = JSON.parse(await readFile(new URL('./node_modules/classnames/package.json', import.meta.url)));
```
These top-level awaits are only supported from Node 14 and up, so we'll need a fix for older versions.

jonkoops commented:
Node 10 doesn't support ES modules, so like expected the tests will fail. Since it will be end-of-life at the end of the month this is not really a problem as we can simply drop support.

jonkoops commented:
@dcousens How would you feel about creating a `next` branch where all breaking PRs like these can be merged? For example, Node 10 support could be dropped there, ES2015 syntax changes could be introduced, etc.

dcousens commented:
@jonkoops I think that's a great idea;  the exact path forward through the mess is a little bit hazy with how `bower` was integrated into the repository,  [there seems to be some concern around breaking the `master` branch](https://github.com/JedWatson/classnames/issues/238#issuecomment-815308016),  but I think inevitably we have to move forward.

A `next` branch works for me.

dcousens commented:
@jonkoops / @remcohaszing thoughts on a typescript codebase?

jonkoops commented:
> @jonkoops / @remcohaszing thoughts on a typescript codebase instead? Our types aren't complicated, but it could simplify the transpilation story.

It should not be a problem, in fact if we add a compilation step we could also maintain compatibility with CommonJS if we wanted to do so.

jonkoops commented:
> A `next` branch works for me.

Great! If you could create one in sync with `master` I will direct any breaking changes there.

remcohaszing commented:
> thoughts on a typescript codebase instead?

I’m all for it, but beware there are some compatibility issues when combining ESM with various tools, such as `ts-node` or `ts-jest`.

---

> It should not be a problem, in fact if we add a compilation step we could also maintain compatibility with CommonJS if we wanted to do so.

TypeScript would compile default exports to:

```js
module.exports.default = classNames;
```

meaning commonjs users would need to use:

```js
const { default: classNames } = require('classnames');
```

Named exports would resolve this issue though.

---

Various big players in the NodeJS ecosystem are pushing forward the philosophy that ESM should be the only module syntax and commonjs should be dropped rather sooner than later. I agree with this philosophy (even though this is currently inconvenient for me as a TypeScript / Jest user).

jonkoops commented:
Remco is right here, this would create a breaking change in the default export for CommonJS. Perhaps it's a good idea to bit the bullet and just move to ESM completely.

dcousens commented:
> Various big players in the NodeJS ecosystem are pushing forward the philosophy that ESM should be the only module syntax and commonjs should be dropped rather sooner than later. I agree with this philosophy (even though this is currently inconvenient for me as a TypeScript / Jest user).

I think, provided we have a stable major,  we can look at these options for the benefits they offer.   I think distributing ESM instead of transpiled JS would be a great benefit.

jonkoops commented:
I've gone ahead and changed the target branch to `next` where we can merge the breaking changes like these. I've added the changes form #252 to ensure that the CI passes properly and I will rebase those changes out once they have been merged to `next`.

jonkoops commented:
I've created #253 to drop support for Bower, as it to my knowledge is not possible to support ESM. Besides this the Bower team actively reccomends switching to NPM or Yarn, so this seems like a good time to bite the bullet.

jonkoops commented on package.json:
```diff
@@ -2,14 +2,24 @@
   },
   "types": "./index.d.ts",
+  "engines": {
+    "node": ">=12.22.0"
```
According to the Node 12 [changelog](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V12.md#2021-03-30-version-12220-erbium-lts-richardlau) ES modules are considered stable since version 12.22.0, so if we were to use this syntax should we not be using `^12.22`? The exact same reference is present in the Node 14 [changelog](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V14.md#2021-05-11-version-14170-fermium-lts-danielleadams), referencing this [commit](https://github.com/nodejs/node/commit/c1af59384d)  for version 14.17.0.

I can't find references to it being stable before that, but this could simply be a docs change and the code could have stabilized before. If these exist could you point them out?

jaydenseric replied:
> I can't find references to it being stable before that, but this could simply be a docs change and the code could have stabilized before.

That's right, marking ESM as "stable" in v12.22.0 was just a docs update:

https://github.com/nodejs/node/pull/35781/files

A rationalization for why I use precisely `^12.20 || >= 14.13` in my own packages would take a lengthy blog post, relating to specific aspects of Node.js ESM and package `exports` functionality, the dev dependencies that I use and their level of support, and the way certain commonjs dependencies are structured (e.g. in tests of my packages doing named imports from a `graphql` peer/dev dependency, I rely on Node.js support for detection of CommonJS named exports that was only introduced in v12.20.0, v14.13.0).

My original warning that the range `>=12.22.0` is incorrect was more referring to it allowing v13.x versions that are not compatible.

Your PR is not using package `exports` subpath patterns, but if it does in the future, or if any dependencies or dev dependencies do (very likely) then you need to account for it only being added in v12.20.0, v14.13.0:

https://nodejs.org/api/packages.html#packages_subpath_patterns

jonkoops replied:
This makes sense to me, I'll add the version range of `^12.20 || >= 14.13` since that seems to be the lowest and still compatible one.

jonkoops commented on package.json:
```diff
@@ -2,14 +2,24 @@
   "main": "index.js",
+  "exports": {
+    ".": "./index.js",
+    "./bind": "./bind.js",
+    "./dedupe": "./dedupe.js"
```
I'm following the examples from the [Node documentation](https://nodejs.org/api/packages.html#packages_subpath_exports) which seem to prefer the style without extensions. I am inclined to follow the extension-less style as it matches the existing API surface and would not require users to update their own code.

Unless I am mistaken tools like UNPKG will still properly resolve the correct filename (such as http://unpkg.com/classnames@latest/dedupe) since we are not changing the root structure of the package. Meaning that since we have `dedupe.js` in the root of the our file structure it can still be resolved by tools that do not support the `exports` field.

There are also alternative tools such as Skypack that are [actively encouraging](https://www.skypack.dev/blog/2020/10/skypack-quality-score-actionable-feedback-to-build-better-packages/) including these fields as well, extensions or not.

jonkoops commented on tests/dedupe.js:
```diff
@@ -1,7 +1,7 @@
 
-var assert = require('assert');
-var dedupe = require('../dedupe');
+import assert from 'assert';
```
Would it also make more sense to use `node:` internal namespace for these built-ins? For example:

```js
import { assert } from 'node:assert'
```

jaydenseric replied:
I think you mean:

```js
import { strictEqual } from 'node:assert';
```

If you do use the `node:` internal namespace, you need to account for it being added in Node.js v12.20.0, v14.13.1 in the package `engines.node` field:

https://nodejs.org/api/esm.html#esm_node_imports

I don't have a hard opinion if the `node:` internal namespace is a good idea for package authors to use yet. Up front it doesn't have a concrete benefit, but it smells like the sort of thing that might mess with people's bundler config and detection of Node.js imports for marking as external, people's regex, etc. Generally it's a good idea though, and @sindresorhus is using it in his packages and encouraging it for the wider ecosystem migrating to ESM:

https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#how-can-i-move-my-commonjs-project-to-esm

jonkoops replied:
Since these files are reserved for the tests we are running and are not shipped in the final package I think we can actually make this change without any impact on users. I've gone ahead and used named exports for all Node internals, including the namespace.

jonkoops commented on package.json:
```diff
@@ -2,14 +2,24 @@
+    ".": "./index.js",
+    "./bind": "./bind.js",
+    "./dedupe": "./dedupe.js"
+  },
```
Do you have examples of tools that would require this kind of change and where it breaks?

jaydenseric replied:
All sorts of tools and userland code. Here is a good entry point to the topic:

https://github.com/nodejs/modules/issues/445

remcohaszing replied:
Just throwing in another option I proposed before: if everything is exported as named exports from `index.js`, no export map is needed at all.

```js
import { bind, classNames, dedupe } from 'classnames';
```

jonkoops replied:
Although the named exports are nice, I think we are losing the semantic value that is provided by the name if the function. 

For example:

```jsx
import { bind } from 'classnames'

<div className={bind('foo, 'bar')}>
```

or:

```jsx
import { bind as classNames } from 'classnames'

<div className={classNames('foo, 'bar')}>
```

This seem undesirable to me as an API surface. Besides that I am sure that there are already applications that are using aliases (for example in [WebPack](https://webpack.js.org/configuration/resolve/) or Typescript compilers) to change the import during compilation.

jaydenseric replied:
@remcohaszing the entire API should be exported from the main index as named exports, but only for tools like  https://bundlephobia.com to be able to access the entire API at once for analysis. Application code should never-ever use a named import from an index if a deep import of the same API is available. Deep imports are far superior, as only the required code loads in memory and bundlers don't need to do any "tree shaking" at all, which is unreliable and resource intense.

jaydenseric replied:
@jonkoops 

> Although the named exports are nice, I think we are losing the semantic value that is provided by the name of the function.

`bind` and `dedupe` are terrible function names in the first place; they don't describe what they do at all. If they were named something more descriptive everything would make sense as named exports in the index and users wouldn't need to alias the imports. The deep import paths would make more sense too if they matched more descriptive function names - I had to look up what "bind" even means in the readme right now.

jonkoops replied:
I think keeping things backwards compatible is enough reason in itself. But let's keep this thread on-topic, in relation to the package file import itself.

jaydenseric commented on benchmarks/run.js:
```diff
@@ -1,17 +1,21 @@
-var local = require('../');
-var dedupe = require('../dedupe');
-var localPackage = require('../package.json');
+import { readFile } from 'fs/promises'
```
This deep `fs/promises` path is only available since Node.js v14.0.0, which is incompatible with the declared Node.js support in the package `engines.node` field.

https://nodejs.org/api/fs.html#fs_promises_api

jonkoops replied:
Since we're not shipping this code to end users and only running it on our CI it seems fine to use this.

jonkoops commented on benchmarks/package.json:
```diff
@@ -4,6 +4,7 @@
   "private": true,
   "description": "",
   "main": "run.js",
+  "type": "module",
```
I know there has been a lot of discussion about this in the past, being a huge topic that split the modules proposal in Node. In my opinion ESM is the future of modules in JS, and for example on the web we are already actively using `.js` in combination with `type="module"`. I do not see the need to change a file extension just to get behavior that in the future should be the default for all packages anyways.

Either way this seem to be more of a personal preference than anything else.

As for the the deep imports, I believe the fact that we use the sub-path exports in the package explicitly makes the 'importable' API surface very clear. It even makes it impossible to make file imports that are not specified in the `exports` field.

jonkoops commented:
I've removed Browserify and replaced it with Rollup, this allows the browser tests to run properly again. I've also added the benchmarks to the CI so they will be run on every pull request.

With all the original functionality now working again I am going ahead and mark this PR ready for review.

jonkoops commented:
@dcousens I see you removed the package lock file, be aware that the CI relies on `npm ci` which requires the lockfile to run.

dcousens commented:
thanks @jonkoops,   I did notice that,  it's more the thought that I'm thinking of merging this as a work in progress sooner than let the branch go stale.
